_blog/2023-10-25-htmx-with-rust.mdx

```diff
@@ -0,0 +1,265 @@
+
+    let router = Router::new()
+        .route("/", get(home))
+        .route("/stream", get(stream))
+        .route("/todos", get(fetch_todos).post(create_todo))
+        .route("/todos/:id", delete(delete_todo))
+        // new handler - we will make this later
+        .route("/todos/stream", get(handle_stream))
+        .with_state(state)
+        .layer(Extension(tx))
+        .layer(Extension(rx));
+
+    Ok(router.into())
+}
+```
+
+Now we want to send a message from our `create_todo` and `delete_todo` handlers through our channel - we can add them by including our extension in the function signature, then after a successful SQL transaction we simply send a message to the channel:
+```rust
+// src/main.rs
+async fn create_todo(
+    State(state): State<AppState>,
+    Extension(tx): Extension<TodosStream>,          
+    Form(form): Form<TodoNew>,
+) -> impl IntoResponse {
+    let todo = sqlx::query_as::<_, Todo>(
+        "INSERT INTO TODOS (description) VALUES ($1) RETURNING id, description",
+    )   
+    .bind(form.description)
+    .fetch_one(&state.db)
+    .await
+    .unwrap();
+
+    if let Err(e) = tx.send( TodoUpdate {mutation_kind: MutationKind::Create, id: todo.id }) {
+      eprintln!("Tried to send log of record with ID {} created but something went wrong: {e}", todo.id);
+     }
+     
+    TodoNewTemplate { todo }
+}
+
+async fn delete_todo(
+State(state): State<AppState>, 
+        Path(id): Path<i32>,
+    Extension(tx): Extension<TodosStream>,          
+) -> impl IntoResponse {
+    sqlx::query("DELETE FROM TODOS WHERE ID = $1")
+        .bind(id)
+        .execute(&state.db)
+        .await
+        .unwrap();
+
+    if let Err(e) = tx.send( TodoUpdate {mutation_kind: MutationKind::Delete, id }) {
+      eprintln!("Tried to send log of record with ID {id} created but something went wrong: {e}");
+    }
+    StatusCode::OK
+}
+```
+
+Now we need to implement the stream handler, which we can do by creating a `BroadcastStream` and then mapping our stream to Axum SSE events:
+
+```rust
+pub async fn handle_stream(
+    Extension(tx): Extension<TodosStream>,
+) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
+    let rx = tx.subscribe();
+
+    let stream = BroadcastStream::new(rx);
+
+// map the stream to axum Events which get sent through the SSE stream
+    Sse::new(
+        stream
+            .map(|msg| {
+                let msg = msg.unwrap();
+                // wrap the message in HTML because htmx expects a HTML fragment response
+                let json = format!("<div>{}</div>", json!(msg));
+                Event::default().data(json)
+            })
+            .map(Ok),
+    )   
+    .keep_alive(
+        axum::response::sse::KeepAlive::new()
+            .interval(Duration::from_secs(600))
+            .text("keep-alive-text"),
+    )   
+}
+```
+Then in our HTML, we will want to add an element that looks like this:
+```html 
+<!-- templates/stream.html -->
```

Comment:
btw, doesn't the mdx here support file names after the language tag?

Reply:
It doesn't show on the Blog page which means it doesn't actually do anything.

Might add these back in though. Not sure